File: spa/src/lib/rut.ts
/**
 * Chilean RUT (Rol Único Tributario) validation and formatting.
 *
 * Valid formats accepted: "12345678-5", "12.345.678-5", "123456785"
 * Stored format (clean): "123456785" (digits + check digit, no separators)
 */

/**
 * Remove dots and dashes, uppercase K.
 */
export function cleanRut(rut: string): string {
  return rut.replace(/[.\-\s]/g, "").toUpperCase();
}

/**
 * Compute the RUT check digit (dígito verificador).
 * Returns "0"-"9" or "K".
 */
export function computeCheckDigit(body: string): string {
  const digits = body.replace(/\D/g, "");
  let sum = 0;
  let mul = 2;
  for (let i = digits.length - 1; i >= 0; i--) {
    sum += parseInt(digits[i], 10) * mul;
    mul = mul === 7 ? 2 : mul + 1;
  }
  const remainder = 11 - (sum % 11);
  if (remainder === 11) return "0";
  if (remainder === 10) return "K";
  return String(remainder);
}

/**
 * Validate a Chilean RUT string.
 * Accepts with or without dots/dashes.
 */
export function isValidRut(rut: string): boolean {
  const clean = cleanRut(rut);
  if (clean.length < 2) return false;

  const body = clean.slice(0, -1);
  const dv = clean.slice(-1);

  // Body must be all digits, 1-8 digits
  if (!/^\d{1,8}$/.test(body)) return false;

  return computeCheckDigit(body) === dv;
}

/**
 * Format a RUT for display: "12.345.678-5"
 */
export function formatRut(rut: string): string {
  const clean = cleanRut(rut);
  if (clean.length < 2) return rut;

  const body = clean.slice(0, -1);
  const dv = clean.slice(-1);

  const formatted = body.replace(/\B(?=(\d{3})+(?!\d))/g, ".");
  return `${formatted}-${dv}`;
}

/**
 * Auto-format on input: adds dash before check digit.
 * Returns the cleaned value (no dots, just digits + optional K).
 */
export function formatRutInput(value: string): string {
  const clean = cleanRut(value);
  if (clean.length <= 1) return clean;

  const body = clean.slice(0, -1).replace(/\D/g, "");
  const dv = clean.slice(-1);
  return `${body}-${dv}`;
}

File: backend/app/rut.py
"""
Chilean RUT (Rol Único Tributario) validation.

Valid formats: "12345678-5", "12.345.678-5", "123456785"
Clean format (stored): "123456785"
"""

import re


def clean_rut(rut: str) -> str:
    """Remove dots, dashes, and spaces; uppercase K."""
    return re.sub(r"[.\-\s]", "", rut).upper()


def compute_check_digit(body: str) -> str:
    """Compute RUT check digit (dígito verificador)."""
    digits = re.sub(r"\D", "", body)
    total = 0
    mul = 2
    for ch in reversed(digits):
        total += int(ch) * mul
        mul = 2 if mul == 7 else mul + 1
    remainder = 11 - (total % 11)
    if remainder == 11:
        return "0"
    if remainder == 10:
        return "K"
    return str(remainder)


def is_valid_rut(rut: str) -> bool:
    """Validate a Chilean RUT string."""
    clean = clean_rut(rut)
    if len(clean) < 2:
        return False

    body = clean[:-1]
    dv = clean[-1]

    if not re.match(r"^\d{1,8}$", body):
        return False

    return compute_check_digit(body) == dv


def format_rut(rut: str) -> str:
    """Format a RUT for display: 12.345.678-5"""
    clean = clean_rut(rut)
    if len(clean) < 2:
        return rut
    body = clean[:-1]
    dv = clean[-1]
    # Add dots every 3 digits from right
    formatted = ""
    for i, ch in enumerate(reversed(body)):
        if i > 0 and i % 3 == 0:
            formatted = "." + formatted
        formatted = ch + formatted
    return f"{formatted}-{dv}"
